daemon/cluster/services.go

```diff
@@ -430,9 +430,17 @@ func (c *Cluster) ServiceLogs(ctx context.Context, selector *backend.LogSelector
 				if err != nil {
 					m.Err = err
 				}
+				// copy over all of the details
+				for _, d := range msg.Attrs {
+					m.Attrs[d.Key] = d.Value
+				}
```

Comment:
Rather than reformat these as a map, would it be possible to change the engine to use a slice internally? I think this would make things cleaner, since we end up sorting the attrs anyway, which involves going back to a slice. The round-trip through a map doesn't seem necessary. See #32418